src/js/utils.js

```diff
@@ -58,6 +58,28 @@ export const drawMetricSummary = (options, client, value, isMedian=true, change=
   }
 };
 
+/**
+ * Invokes the callback when the element is visible for the first time.
+ * 
+ * Rendering all charts immediately at startup causes long tasks and slow INP.
+ * Instead, wait until the chart is in view before rendering.
+ * Rendering a single chart might still create a long task,
+ * but it'll be less likely to contribute to INP.
+ * 
+ * @param {HTMLElement} element 
+ * @param {Function} callback 
+ */
+export function callOnceWhenVisible(element, callback) {
+  new IntersectionObserver((entries, observer) => {
+    if (!entries[0].isIntersecting) {
+      return;
+    }
+
+    observer.unobserve(element);
+    callback();
+  }).observe(element);
```

Comment:
What about adding a threshold so charts are loaded when just off screen?

```suggestion
  }, {rootMargin: "1000px"}).observe(element);
```

Reply:
That'd make it more likely for the first chart to load at startup and cause issues with early interactions. Not sure the UX is worth the INP tradeoff. WDYT?

Reply:
I was assuming that one chart wouldn't cause that much of an issue and it's the combination of doing so many charts at once that's the problem. Particularly for CrUX report which has 17 charts.

Personally I feel that lazy loading content (which effectively is what we're doing here - though it's more "lazy rendering") should be as invisible to the user as possible. Not rendering ANYTHING feels like we're optimising for the metric than the user. It's also the reason why `loading="lazy"` has thresholds to try to render some content as it approaches the viewport. So yeah I'd argue the UX is worth the INP trade off.

Plus, when the user clicks any chart, we scroll down and have to render ALL the graphs above it. At least getting the first one out of the way is one less to render in that case. Though, on the flip side, we'll likely render one more chart after the one we want too so there might not be any net gain, except when scrolling to the end (though the later one should at least render after the one being scrolled to).

We could try that and monitor INP scores? If this is enough to fix it then leave it in (or maybe even increase it!). If not, then can remove the threshold and go with 0.

Reply:
Saw your LGTM comment, thanks.

> Not rendering ANYTHING feels like we're optimising for the metric than the user.

Just wanted to add that this is definitely my biggest concern with this approach. I think it's mitigated in part by the fact that rendering a single chart takes ~400ms (4x throttling on an M1), and IMO it's a more tolerable UX to wait that long to see the chart than it would be to wait that long after an interaction. And as a small consolation, rendering the chart when it's visible lets the user see the fancy animation of the data taking shape.

>Plus, when the user clicks any chart, we scroll down and have to render ALL the graphs above it. 

I tested this by jumping to a chart in the middle of the page and I'm only seeing that chart render, not the ones above it. Maybe that behavior is unique to IntersectionObserver, unlike `loading=lazy`?

Anyway, with your LGTM I'll resolve this and open a new PR to help with A/B testing the change.

Reply:
> I tested this by jumping to a chart in the middle of the page and I'm only seeing that chart render, not the ones above it.

Ohhhh it was only rendering all the charts above with my suggested change! So maybe that was a good reason not to do that!